<launch>
    <!-- unique camera name-->
    <arg name="camera_name" default="camera"/>
    <!-- Hardware depth registration -->
    <arg name="depth_registration" default="true"/>
    <arg name="serial_number" default=""/>
    <arg name="device_num" default="1"/>
    <arg name="vendor_id" default="0x2bc5"/>
    <arg name="product_id" default=""/>

    <arg name="enable_point_cloud" default="false"/>
    <arg name="enable_colored_point_cloud" default="true"/>
    <arg name="connection_delay" default="100"/>
    <arg name="color_width" default="640"/>
    <arg name="color_height" default="480"/>
    <arg name="color_fps" default="30"/>
    <arg name="enable_color" default="true"/>
    <arg name="color_format" default="MJPG"/>
    <arg name="flip_color" default="false"/>
    <arg name="enable_color_auto_exposure" default="true"/>
    <arg name="depth_width" default="640"/>
    <arg name="depth_height" default="400"/>
    <arg name="depth_fps" default="30"/>
    <arg name="enable_depth" default="true"/>
    <arg name="depth_format" default="Y16"/>
    <arg name="flip_depth" default="false"/>
    <arg name="depth_precision" default="0.4mm"/>
    <arg name="ir_width" default="640"/>
    <arg name="ir_height" default="400"/>
    <arg name="ir_fps" default="30"/>
    <arg name="enable_ir" default="false"/>
    <arg name="ir_format" default="Y8"/>
    <arg name="flip_ir" default="false"/>
    <arg name="enable_ir_auto_exposure" default="true"/>
    <arg name="publish_tf" default="true"/>
    <arg name="tf_publish_rate" default="10.0"/>
    <arg name="ir_info_uri" default=""/>
    <arg name="color_info_uri" default=""/>
    <arg name="log_level" default="none"/>
    <arg name="enable_d2c_viewer" default="false"/>
    <arg name="enable_pipeline" default="true"/>
    <arg name="enable_ldp" default="true"/>
    <arg name="enable_soft_filter" default="true"/>
    <arg name="soft_filter_max_diff" default="-1"/>
    <arg name="soft_filter_speckle_size" default="-1"/>
    <arg name="sync_mode" default="close"/>
    <arg name="ir_trigger_signal_in_delay" default="0"/>
    <arg name="rgb_trigger_signal_in_delay" default="0"/>
    <arg name="device_trigger_signal_out_delay" default="0"/>
    <arg name="trigger_signal_out" default="false"/>

     <arg name="enable_laser" default="true"/>
     <arg name="floor_height" default="0.24"/>

     <!-- Khai báo các tham số -->
  <arg name="pass_through_x_min" default="-1.5" />
  <arg name="pass_through_x_max" default="1.5" />
  <arg name="pass_through_y_min" default="-1.5" />
  <arg name="pass_through_y_max" default="0.12" />
  <arg name="pass_through_z_min" default="0.0" />
  <arg name="pass_through_z_max" default="4.0" />


  <arg name="voxel_leaf_size" default="0.01" />

  <arg name="ground_seg_max_iterations" default="1" />
  <arg name="ground_seg_distance_threshold" default="0.01" />

  <arg name="cluster_tolerance" default="0.02" />
  <arg name="min_cluster_size" default="100" />
  <arg name="max_cluster_size" default="25000" />

  <arg name="safety_warn_size" default="2.2" />
  <arg name="safety_protect_size" default="1.2" />
  <arg name="safety_warn_position_x" default="0.0" />
  <arg name="safety_warn_position_y" default="0.0" />
  <arg name="safety_warn_position_z" default="1.1" />
  <arg name="safety_protect_position_x" default="0.0" />
  <arg name="safety_protect_position_y" default="0.0" />
  <arg name="safety_protect_position_z" default="0.6" />

  <arg name="min_cluster_warn_size" default="50" />
  <arg name="min_cluster_protect_size" default="50" />
  <arg name="min_consecutive_warn_count" default="5" />
  <arg name="min_consecutive_protect_count" default="5" />


    <group ns="$(arg camera_name)">
        <node name="camera" pkg="orbbec_camera" type="orbbec_camera_node" output="screen">
            <param name="camera_name" value="$(arg camera_name)"/>
            <param name="depth_registration" value="$(arg depth_registration)"/>
            <param name="serial_number" type="string" value="$(arg serial_number)"/>
            <param name="vendor_id" value="$(arg vendor_id)"/>
            <param name="product_id" value="$(arg product_id)"/>
            <param name="enable_point_cloud" value="$(arg enable_point_cloud)"/>
            <param name="enable_colored_point_cloud" value="$(arg enable_colored_point_cloud)"/>
            <param name="connection_delay" value="$(arg connection_delay)"/>
            <param name="color_width" value="$(arg color_width)"/>
            <param name="color_height" value="$(arg color_height)"/>
            <param name="color_fps" value="$(arg color_fps)"/>
            <param name="enable_color" value="$(arg enable_color)"/>
            <param name="color_format" value="$(arg color_format)"/>
            <param name="flip_color" value="$(arg flip_color)"/>
            <param name="enable_color_auto_exposure" value="$(arg enable_color_auto_exposure)"/>
            <param name="depth_width" value="$(arg depth_width)"/>
            <param name="depth_height" value="$(arg depth_height)"/>
            <param name="depth_fps" value="$(arg depth_fps)"/>
            <param name="enable_depth" value="$(arg enable_depth)"/>
            <param name="depth_format" value="$(arg depth_format)"/>
            <param name="flip_depth" value="$(arg flip_depth)"/>
            <param name="depth_precision" value="$(arg depth_precision)"/>
            <param name="ir_width" value="$(arg ir_width)"/>
            <param name="ir_height" value="$(arg ir_height)"/>
            <param name="ir_fps" value="$(arg ir_fps)"/>
            <param name="enable_ir" value="$(arg enable_ir)"/>
            <param name="ir_format" value="$(arg ir_format)"/>
            <param name="flip_ir" value="$(arg flip_ir)"/>
            <param name="enable_ir_auto_exposure" value="$(arg enable_ir_auto_exposure)"/>
            <param name="publish_tf" value="$(arg publish_tf)"/>
            <param name="tf_publish_rate" value="$(arg tf_publish_rate)"/>
            <param name="ir_info_uri" value="$(arg ir_info_uri)"/>
            <param name="color_info_uri" value="$(arg color_info_uri)"/>
            <param name="log_level" value="$(arg log_level)"/>
            <param name="enable_d2c_viewer" value="$(arg enable_d2c_viewer)"/>
            <param name="enable_pipeline" value="$(arg enable_pipeline)"/>
            <param name="enable_ldp" value="$(arg enable_ldp)"/>
            <param name="device_num" value="$(arg device_num)"/>
            <param name="enable_soft_filter" value="$(arg enable_soft_filter)"/>
            <param name="soft_filter_max_diff" value="$(arg soft_filter_max_diff)"/>
            <param name="soft_filter_speckle_size" value="$(arg soft_filter_speckle_size)"/>
            <param name="sync_mode" value="$(arg sync_mode)"/>
            <param name="ir_trigger_signal_in_delay" value="$(arg ir_trigger_signal_in_delay)"/>
            <param name="rgb_trigger_signal_in_delay" value="$(arg rgb_trigger_signal_in_delay)"/>
            <param name="device_trigger_signal_out_delay" value="$(arg device_trigger_signal_out_delay)"/>
            <param name="trigger_signal_out" value="$(arg trigger_signal_out)"/>

            <param name="enable_laser" value="$(arg enable_laser)"/>
            <param name="floor_height" value="$(arg floor_height)"/>
            <remap from="/$(arg camera_name)/depth_registered/points" to="/$(arg camera_name)/depth/color/points"/>
            
        </node>
    </group>
     <node name="obstacle_detection_node" pkg="obstacle_detection" type="obstacle_detection" output="screen">
    <!-- Truyền các tham số vào node -->
    <param name="pass_through_x_min" value="$(arg pass_through_x_min)" />
    <param name="pass_through_x_max" value="$(arg pass_through_x_max)" />
    <param name="pass_through_y_min" value="$(arg pass_through_y_min)" />
    <param name="pass_through_y_max" value="$(arg pass_through_y_max)" />
    <param name="pass_through_z_min" value="$(arg pass_through_z_min)" />
    <param name="pass_through_z_max" value="$(arg pass_through_z_max)" />

    <param name="voxel_leaf_size" value="$(arg voxel_leaf_size)" />

    <param name="ground_seg_max_iterations" value="$(arg ground_seg_max_iterations)" />
    <param name="ground_seg_distance_threshold" value="$(arg ground_seg_distance_threshold)" />

    <param name="cluster_tolerance" value="$(arg cluster_tolerance)" />
    <param name="min_cluster_size" value="$(arg min_cluster_size)" />
    <param name="max_cluster_size" value="$(arg max_cluster_size)" />

    <param name="safety_warn_size" value="$(arg safety_warn_size)" />
    <param name="safety_protect_size" value="$(arg safety_protect_size)" />
    <param name="safety_warn_position_x" value="$(arg safety_warn_position_x)" />
    <param name="safety_warn_position_y" value="$(arg safety_warn_position_y)" />
    <param name="safety_warn_position_z" value="$(arg safety_warn_position_z)" />
    <param name="safety_protect_position_x" value="$(arg safety_protect_position_x)" />
    <param name="safety_protect_position_y" value="$(arg safety_protect_position_y)" />
    <param name="safety_protect_position_z" value="$(arg safety_protect_position_z)" />

    <param name="min_cluster_warn_size" value="$(arg min_cluster_warn_size)" />
    <param name="min_cluster_protect_size" value="$(arg min_cluster_protect_size)" />
    <param name="min_consecutive_warn_count" value="$(arg min_consecutive_warn_count)" />
    <param name="min_consecutive_protect_count" value="$(arg min_consecutive_protect_count)" />
    <remap from="/obstacle_detection/safety_status" to="/depth_camera/fields_safety"/>
  </node>
  <node pkg="tf2_ros" type="static_transform_publisher" name="camera_tf" args="0 0 0 0 0 0 base_camera camera_link" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="camera_base_tf" args="0.551 0 0.22 -1.5708 0 -1.5708 base_link base_camera" />
  <!-- <node pkg="tf2_ros" type="static_transform_publisher" name="camera_base_tf" args="0.551 0 0.3 -1.5708 0 -1.5708 base_link base_camera" /> -->

</launch>
